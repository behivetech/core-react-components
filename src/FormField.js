// Vendor Libs
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

// Components
import {FormContext} from './Form';
import TextField from './TextField';

export default class FormField extends Component {
    
    componentWillUnmount() {
        this.setFieldValueDebounced.cancel()
    }
    
    handleChange = (event) => {
        const {name, onClick} = this.props;

        this.value = event.target.value;
        this.setFieldValueDebounced(name, event.target.value);
        
        if (onClick) {
            onClick(event);
        }
    }

    renderChild = (formState) => {
        const {className, component, name} = this.props;
        let props = {
            className: classnames('form-text-field', className),
            onChange: this.handleChange,
        }
        let content = null;

        this.value = formState.fieldValues[name] || '';

        if (!this.setFieldValueDebounced) {
            this.setFieldValueDebounced = formState.setFieldValueDebounced;
        } 

        import(`./${component}`).then((Component) => {
            console.log(Component)
            content = <Component {...this.props} {...props} value={this.value} />;
        });

        return content;
    }

    render() {
        return (
            <FormContext.Consumer>
                {this.renderChild}
            </FormContext.Consumer>
        );
    }
}

FormField.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    component: PropTypes.oneOf(['TextField']).isRequired,
    name: PropTypes.string.isRequired,
};