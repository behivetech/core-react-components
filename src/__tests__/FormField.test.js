import FormField from "../FormField";

describe('FormField', () => {
    const getEvent = (target = {}) => {
        return {
            stopPropagation: jest.fn(),
            target: {
                checked: target.checked,
                value: target.value || 'mock event value',
            },
        };
    };
    const mockSetFieldValue = jest.fn();

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render with the correct props', () => {
        const wrapper = mount(<FormField name="mockName" componentName="TextField" />);

        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should set the right value when componentName is Checkbox', () => {
         const wrapperProps = {
            name: 'mockName',
            componentName: 'Checkbox',
            defaultValue: 'mock default value',
        };
        const wrapperInstance = mount(<FormField {...wrapperProps} />).instance();
        const mockEvent = getEvent({checked: true});

        wrapperInstance.setFieldValue = mockSetFieldValue;
        wrapperInstance.handleClick(mockEvent);
        expect(mockSetFieldValue).toBeCalledWith('mockName', wrapperProps.defaultValue);
    });

    it('should unset the right values when componentName is Checkbox', () => {
        const wrapperProps = {
            name: 'mockName',
            componentName: 'Checkbox',
            defaultValue: 'mock default value',
        };
        const wrapperInstance = mount(<FormField {...wrapperProps} />).instance();
        const mockEvent = getEvent({checked: false});

        wrapperInstance.setFieldValue = mockSetFieldValue;
        wrapperInstance.handleClick(mockEvent);
        expect(mockSetFieldValue).toBeCalledWith('mockName', "");
    });

    it('should set the right values when componentName is Checkbox and value is array', () => {
        const wrapperProps = {
            name: 'mockName[]',
            componentName: 'Checkbox',
            defaultValue: 'mock default value',
            value: ['mock value 1'],
        };
        const wrapperInstance = mount(<FormField {...wrapperProps} />).instance();
        const mockEvent = getEvent({checked: true});

        wrapperInstance.setFieldValue = mockSetFieldValue;
        wrapperInstance.getFieldValue = () => wrapperProps.value;
        wrapperInstance.handleClick(mockEvent);
        expect(mockSetFieldValue).toBeCalledWith(
            'mockName',
            [...wrapperProps.value, ...[wrapperProps.defaultValue]]
        );
    });

    it('should unset the right values when componentName is Checkbox and value is array', () => {
        const wrapperProps = {
            name: 'mockName[]',
            componentName: 'Checkbox',
            defaultValue: 'mock default value',
            value: ['mock value 1'],
        };
        const wrapperInstance = mount(<FormField {...wrapperProps} />).instance();
        const mockEvent = getEvent({checked: false});

        wrapperInstance.setFieldValue = mockSetFieldValue;
        wrapperInstance.getFieldValue = () => [...wrapperProps.value, ...[wrapperProps.defaultValue]];
        wrapperInstance.handleClick(mockEvent);
        expect(mockSetFieldValue).toBeCalledWith('mockName', wrapperProps.value);
    });

    it('should set the value to false when componentName is Checkbox and defaultValue is true', () => {
        const wrapperProps = {
            name: 'mockName',
            componentName: 'Checkbox',
            defaultValue: true,
        };
        const wrapperInstance = mount(<FormField {...wrapperProps} />).instance();
        const mockEvent = getEvent({checked: false});

        wrapperInstance.setFieldValue = mockSetFieldValue;
        wrapperInstance.formStateValue = true;
        wrapperInstance.handleClick(mockEvent);
        expect(mockSetFieldValue).toBeCalledWith('mockName', false);
    });
});

