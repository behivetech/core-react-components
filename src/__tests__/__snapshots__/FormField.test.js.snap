// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should render a child 1`] = `"<FormField name=\\"mockName\\" componentName=\\"TextField\\" validate={{...}} />"`;

exports[`FormField should render with the correct props 1`] = `"<FormField name=\\"mockName\\" componentName=\\"TextField\\" />"`;
